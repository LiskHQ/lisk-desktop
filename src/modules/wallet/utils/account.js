/* eslint-disable max-lines, max-len */
import { passphrase as LiskPassphrase, cryptography } from '@liskhq/lisk-client';
import { regex } from 'src/const/regex';

/**
 * Extracts Lisk PrivateKey/PublicKey pair from a given valid Mnemonic passphrase
 *
 * @param {String} passphrase - Valid Mnemonic passphrase
 * @param {boolean} enableCustomDerivationPath - enable custom derivation for HW
 * @param {String} derivationPath - custom derivation path for HW
 * @returns {object} - Extracted publicKey for a given valid passphrase
 */
export const extractKeyPair = async ({
  passphrase, enableCustomDerivationPath = false, derivationPath,
}) => {
  if (enableCustomDerivationPath) {
    const privateKey = await cryptography.ed.getKeyPairFromPhraseAndPath(passphrase, derivationPath);
    const publicKey = cryptography.ed.getPublicKeyFromPrivateKey(privateKey).toString('hex');
    return {
      publicKey,
      privateKey: privateKey.toString('hex'),
      isValid: true,
    };
  }

  if (LiskPassphrase.Mnemonic.validateMnemonic(passphrase)) {
    const keyPair = cryptography.legacy.getKeys(passphrase);
    return {
      publicKey: keyPair.publicKey.toString('hex'),
      privateKey: keyPair.privateKey.toString('hex'),
      isValid: true,
    };
  }
  return { isValid: false };
};

/**
 * Extracts Lisk PublicKey from a given valid Mnemonic passphrase
 *
 * @param {String} passphrase - Valid Mnemonic passphrase
 * @param {boolean} enableCustomDerivationPath - enable custom derivation for HW
 * @param {String} derivationPath - custom derivation path for HW
 * @returns {String?} - Extracted publicKey for a given valid passphrase
 */
export const extractPublicKey = async (
  passphrase, enableCustomDerivationPath = false, derivationPath,
) => {
  const keyPair = await extractKeyPair({ passphrase, enableCustomDerivationPath, derivationPath });

  if (keyPair.isValid) {
    return keyPair.publicKey;
  }

  throw Error('Invalid passphrase');
};

/**
 * Extracts Lisk PrivateKey from a given valid Mnemonic passphrase
 *
 * @param {String} passphrase - Valid Mnemonic passphrase
 * @param {boolean} enableCustomDerivationPath - enable custom derivation for HW
 * @param {String} derivationPath - custom derivation path for HW
 * @returns {String?} - Extracted PrivateKey for a given valid passphrase
 */
export const extractPrivateKey = async (
  passphrase, enableCustomDerivationPath = false, derivationPath,
) => {
  const keyPair = await extractKeyPair({ passphrase, enableCustomDerivationPath, derivationPath });
  if (keyPair.isValid) {
    return keyPair.privateKey;
  }

  throw Error('Invalid passphrase');
};

/**
 * Extracts address from publicKey
 *
 * @param {String} data PublicKey in Hex
 * @returns {String} - address derived from the given publicKey
 */
export const extractAddressFromPublicKey = (publicKey) => {
  if (regex.publicKey.test(publicKey)) {
    return cryptography.address.getLisk32AddressFromPublicKey(Buffer.from(publicKey, 'hex')).toString('hex');
  }
  if (Buffer.isBuffer(publicKey)) {
    return cryptography.address.getLisk32AddressFromPublicKey(publicKey);
  }
  throw Error(`Unable to convert publicKey ${publicKey} to address`);
};

/**
 * Extracts address from Mnemonic passphrase
 *
 * @param {String} data Valid Mnemonic passphrase
 * @returns {String} - address derived from the given passphrase
 */
export const extractAddressFromPassphrase = (data) => {
  if (LiskPassphrase.Mnemonic.validateMnemonic(data)) {
    const { publicKey } = cryptography.legacy.getKeys(data);
    return cryptography.address.getLisk32AddressFromPublicKey(publicKey).toString('hex');
  }
  throw Error('Invalid passphrase');
};

/**
 * Extracts Lisk address from given passphrase or publicKey
 *
 * @param {String} data - passphrase or public key
 * @returns {String?} - Extracted address for a given valid passphrase or publicKey
 */
export const getBase32AddressFromAddress = (data) => {
  try {
    return cryptography.address.getLisk32AddressFromAddress(data);
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error(e);
    throw Error('Invalid address');
  }
};

export const getAddressFromBase32Address = (data) => {
  try {
    return cryptography.address.getAddressFromLisk32Address(data);
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error(e);
    throw Error('Invalid address');
  }
};

/**
 * Returns a shorter version of a given address
 * by replacing characters by ellipsis except for
 * the first and last 3.
 * @param {String} address LSk address
 * @param {String?} size An option of small and medium
 * @returns {String} Truncated address
 */
export const truncateAddress = (address, size) => {
  const truncateOptions = ['small', 'medium'];
  const selectedSize = truncateOptions.includes(size) ? size : truncateOptions[0];
  if (!address) return address;
  return address.replace(regex.truncate[selectedSize], '$1...$3');
};

/**
 * calculates the balance locked in votes
 *
 * @param {Object} votes - Votes dictionary, values must include vote.confirmed
 * @returns {Number} - Sum of vote amounts
 */
export const calculateBalanceLockedInVotes = (votes = {}) =>
  Object.values(votes).reduce((total, vote) => (total + vote.confirmed), 0);

/**
 * calculates balance locked for the account in unvotes
 *
 * @param {Array} unlocking - unlocking values array from the account details
 * @returns {Number} - Sum of locked LSK, this can be different than sum of vote amounts
 */
export const calculateBalanceLockedInUnvotes = (unlocking = []) =>
  unlocking.reduce((acc, vote) => acc + parseInt(vote.amount, 10), 0);

/**
 * Checks if given unlocking item can be unlocked
 * (Checks if the current height is greater than the unlocking height)
 *
 * @param {Number} unlockHeight - The height at which given LSK can be unlocked
 * @param {Number} currentBlockHeight - Current block height
 * @returns {Boolean} - True if the height is there
 */
export const isBlockHeightReached = (unlockHeight, currentBlockHeight) =>
  currentBlockHeight >= unlockHeight;

/**
 * returns unlocking objects for broadcasting an unlock transaction
 * at the current height
 *
 * @param {Array} unlocking - unlocking values array from the account details
 * @param {Number} currentBlockHeight - Current block height
 * @returns {Array} Array of LSK rows available to unlock
 */
export const getUnlockableUnlockObjects = (unlocking = [], currentBlockHeight = 0) =>
  unlocking.filter(vote => isBlockHeightReached(vote.height.end, currentBlockHeight))
    .map(vote => ({
      delegateAddress: vote.delegateAddress,
      amount: vote.amount,
      unvoteHeight: Number(vote.height.start),
    }));

/**
 * returns the balance that can be unlocked at the current block height
 *
 * @param {Array} unlocking - unlocking values array from the account details
 * @param {Number} currentBlockHeight - Current block height
 * @returns {Number} - The LSK value that can be unlocked
 */
export const calculateUnlockableBalance = (unlocking = [], currentBlockHeight = 0) =>
  unlocking.reduce(
    (sum, vote) =>
      (isBlockHeightReached(vote.height.end, currentBlockHeight)
        ? sum + parseInt(vote.amount, 10) : sum),
    0,
  );

/**
 * returns the balance that can not be unlocked at the current block height
 *
 * @param {Array} unlocking - unlocking values array from the account details
 * @param {Number} currentBlockHeight - Current block height
 * @returns {Number} - The LSK value that can NOT be unlocked at the current height
 */
export const calculateBalanceUnlockableInTheFuture = (unlocking = [], currentBlockHeight = 0) =>
  unlocking.reduce(
    (sum, vote) =>
      (!isBlockHeightReached(vote.height.end, currentBlockHeight)
        ? sum + parseInt(vote.amount, 10) : sum),
    0,
  );

export const calculateRemainingAndSignedMembers = (
  keys = { optionalKeys: [], mandatoryKeys: [] },
  signaturesInTransaction = [],
  ignoreFirstSignature = false,
) => {
  const signatures = ignoreFirstSignature
    ? signaturesInTransaction.slice(1) : signaturesInTransaction;
  const { mandatoryKeys, optionalKeys } = keys;
  const signed = [];
  const remaining = [];

  mandatoryKeys.forEach((key, index) => {
    const hasSigned = Boolean(signatures[index]);
    const value = {
      publicKey: key, mandatory: true, address: extractAddressFromPublicKey(key),
    };

    if (hasSigned) {
      signed.push(value);
    } else {
      remaining.push(value);
    }
  });

  optionalKeys.forEach((key, index) => {
    const hasSigned = Boolean(signatures[index + mandatoryKeys.length]);
    const value = {
      publicKey: key, mandatory: false, address: extractAddressFromPublicKey(key),
    };

    if (hasSigned) {
      signed.push(value);
    } else {
      remaining.push(value);
    }
  });

  return { signed, remaining };
};

/**
 * Get keys object from account info or multisig tx asset
 * @param {object} data
 * @param {object} data.senderAccount - Account info
 * @param {object} data.transaction - Transaction details
 * @param {boolean} data.isGroupRegistration - tx moduleAsset check
 * @returns {object} - Keys, including number and list of mandatory and optional keys
 */
export const getKeys = ({ senderAccount, transaction, isGroupRegistration }) => {
  if (isGroupRegistration) {
    return transaction.params;
  }

  return senderAccount.keys;
};

export const validate2ndPass = async (account, passphrase, error) => {
  const messages = [];
  if (error) {
    messages.push(messages);
    return messages;
  }

  const secondPublicKey = account.keys.mandatoryKeys
    .filter(item => item !== account.summary.publicKey);
  const publicKey = await extractPublicKey(passphrase);
  // compare them
  if (!secondPublicKey.length || publicKey !== secondPublicKey[0]) {
    messages.push('This passphrase does not belong to your account.');
  }
  return messages;
};
