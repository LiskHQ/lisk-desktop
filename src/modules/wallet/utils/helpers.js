export const calculateSupply = (balance, totalSupply) =>
  Number((balance * 100) / totalSupply).toFixed(2);
