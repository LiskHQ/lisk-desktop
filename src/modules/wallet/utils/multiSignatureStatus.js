import { signatureCollectionStatus } from 'src/modules/transaction/configuration/txStatus';
import {
  getTransactionSignatureStatus,
  showSignButton,
} from '../components/signMultisigView/helpers';

/**
 * Get required statuses of a multi-signature transction
 *
 * @param {Object} data - The object containing, senderAccount, account, transactionJSON and currentAccount
 * @param {Object} data.senderAccount - sender's account
 * @param {Object} data.currentAccount - the currently selected encrypted account
 * @param {Object} daqta.transactionJSON - JSON formate of the transaction to be signed
 * @param {Object} daqta.account - user's account from redux
 * @returns {Object} padded string
 */
export const getMultiSignatureStatus = ({
  senderAccount,
  account,
  transactionJSON,
  currentAccount,
}) => {
  const isMember = showSignButton(senderAccount, account, transactionJSON);
  const signatureStatus = getTransactionSignatureStatus(senderAccount, transactionJSON);
  const canSenderSignTx =
    transactionJSON.senderPublicKey === currentAccount.metadata.pubkey &&
    signatureStatus === signatureCollectionStatus.fullySigned;

  return { isMember, signatureStatus, canSenderSignTx };
};
