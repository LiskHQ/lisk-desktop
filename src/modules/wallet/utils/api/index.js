import { regex } from 'src/const/regex';
import { tokenMap } from '@token/fungible/consts/tokens';
import { HTTP_CODES, HTTP_PREFIX } from 'src/const/httpCodes';
import http from 'src/utils/api/http';
import ws from 'src/utils/api/ws';
import { isEmpty } from 'src/utils/helpers';
import { extractAddressFromPublicKey, extractPublicKey } from '@wallet/utils/account';

const httpPaths = {
  account: `${HTTP_PREFIX}/auth`,
  accounts: `${HTTP_PREFIX}/auth`,
};

const wsMethods = {
  accounts: 'get.accounts',
  account: 'get.accounts',
};

/**
 * Prioritizes the API params and if passed, converts
 * publicKey and passphrase to address.
 *
 * @param {Object} params
 * @param {String?} params.username Valid validator username
 * @param {String?} params.address Valid Lisk Address
 * @param {String?} params.passphrase Valid Mnemonic passphrase
 * @param {String?} params.publicKey Valid Lisk PublicKey
 *
 * @returns {Object} Params containing either address or username
 */
const getAccountParams = async (params) => {
  if (!params || isEmpty(params)) return {};
  const {
    username,
    address,
    passphrase,
    publicKey,
  } = params;

  if (username) return { username };
  if (publicKey) return { publicKey };
  if (address) return { address };
  if (passphrase) {
    return { publicKey: await extractPublicKey(passphrase) };
  }

  return {};
};

/**
 * Retrieves details of an account with given params
 *
 * @param {Object} data
 * @param {Object} data.network The network config from the Redux store
 * @param {String?} data.baseUrl Custom API URL
 * @param {Object} data.params
 * @param {String?} data.params.username Valid validator username
 * @param {String?} data.params.address Valid Lisk Address
 * @param {String?} data.params.passphrase Valid Mnemonic passphrase
 * @param {String?} data.params.publicKey Valid Lisk PublicKey
 *
 * @returns {Promise}
 */
// eslint-disable-next-line complexity, max-statements
export const getAccount = async ({
  network, params, baseUrl,
}) => {
  const normParams = await getAccountParams(params);

  try {
    const response = await http({
      baseUrl,
      path: httpPaths.account,
      network,
      params: normParams,
    });

    if (response.data) {
      const account = {
        keys: { ...response.data },
        publicKey: response.meta?.publicKey ?? '',
      };
      return account;
    }
  } catch (e) {
    if (e.code === HTTP_CODES.NOT_FOUND) {
      // eslint-disable-next-line no-console
      console.log('Lisk account not found.');

      if (params.publicKey) {
        const address = extractAddressFromPublicKey(params.publicKey);
        const account = {
          summary: {
            publicKey: params.publicKey,
            privateKey: params.privateKey,
            balance: 0,
            address,
            token: tokenMap.LSK.key,
          },
          sequence: {
            nonce: 0,
          },
        };
        return account;
      }
    } else {
      throw Error(e);
    }
  }
  throw Error('Error retrieving account');
};

const accountFilters = {
  limit: { key: 'limit', test: num => (typeof num === 'number') },
  offset: { key: 'offset', test: num => (typeof num === 'number' && num > 0) },
  sort: {
    key: 'sort',
    test: str => ['balance:asc', 'balance:desc'].includes(str),
  },
};

const getRequests = (values, isDelegate) => {
  const paramList = values.find(item => Array.isArray(item.list) && item.list.length);
  if (paramList) {
    return paramList.list
      .filter(item => regex[paramList.name].test(item))
      .map((item) => {
        const params = isDelegate
          ? {
            [paramList.name]: item,
            isDelegate: true,
          } : {
            [paramList.name]: item,
          };
        return {
          method: wsMethods.accounts,
          params,
          jsonrpc: '2.0',
        };
      });
  }
  return false;
};

/**
 * Retrieves the list of accounts with given params
 *
 * @param {Object} data
 * @param {Object} data.network The network config from the Redux store
 * @param {String?} data.baseUrl Custom API URL
 * @param {Object} data.params
 * @param {String?} data.params.usernameList Valid validator username
 * @param {String?} data.params.addressList Valid Lisk Address
 * @param {String?} data.params.publicKeyList Valid Lisk PublicKey
 * @param {String?} data.params.limit Used for pagination
 * @param {String?} data.params.offset Used for pagination
 * @param {String?} data.params.sort  an option of 'balance:asc' and 'balance:desc',
 *
 * @returns {Promise}
 */
export const getAccounts = async ({
  network,
  params = {},
  baseUrl,
}) => {
  // Use websocket to retrieve accounts with a given array of addresses
  const requests = getRequests([
    { name: 'address', list: params.addressList },
    { name: 'publicKey', list: params.publicKeyList },
    { name: 'username', list: params.usernameList },
  ], params.isDelegate);
  if (requests.length) {
    return ws({
      requests,
      baseUrl: baseUrl || network.networks.LSK.serviceUrl,
    });
  }

  // Use HTTP to retrieve accounts with given sorting and pagination parameters
  const normParams = {};
  Object.keys(params).forEach((key) => {
    if (accountFilters[key] && accountFilters[key].test(params[key])) {
      normParams[accountFilters[key].key] = params[key];
    }
  });

  return http({
    path: httpPaths.accounts,
    network,
    baseUrl,
    params: normParams,
  });
};
