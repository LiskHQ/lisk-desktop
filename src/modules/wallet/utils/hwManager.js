// eslint-disable-next-line import/no-unresolved
import i18next from 'i18next';
import {
  checkIfInsideLiskApp,
  getAddress,
  getPublicKey,
  subscribeToDeviceConnected,
  subscribeToDeviceDisconnected,
  subscribeToDevicesList,
  validatePin,
  signMessage,
} from '@libs/hwServer/communication';
import { extractAddressFromPublicKey } from './account';
import { getUsedHWAccounts } from './api';

const getAccountBundle = async (deviceId, network, offset) => {
  const publicKeyList = [];

  for (let index = offset; index < offset + 3; index++) {
    // eslint-disable-next-line no-await-in-loop
    const publicKey = await getPublicKey({ index, deviceId });
    publicKeyList.push(publicKey);
  }
  const accounts = await getUsedHWAccounts(publicKeyList);
  return accounts.filter(item => item.availableBalance);
};

/**
 * getAccountsFromDevice - Function.
 * This function is used for retrieve the accounts from an hw device, using public keys.
 */
const getAccountsFromDevice = async ({ device: { deviceId }, network }) => {
  let accounts = [];

  for (let index = 0; index <= accounts.length; index += 10) {
    // eslint-disable-next-line no-await-in-loop
    const result = await getAccountBundle(deviceId, network, index);
    // eslint-disable-next-line no-await-in-loop
    accounts = [...accounts, ...result];
  }
  return accounts;
};

const getNewAccountByIndex = async ({ device: { deviceId }, index }) => {
  const publicKey = await getPublicKey({ index, deviceId });

  return {
    summary: {
      publicKey,
      address: extractAddressFromPublicKey(publicKey),
      balance: '0',
    },
  };
};

const isKeyMatch = (aPublicKey, signerPublicKey) => (Buffer.isBuffer(aPublicKey)
  ? aPublicKey.equals(signerPublicKey)
  : Buffer.from(aPublicKey, 'hex').equals(signerPublicKey));

/**
 * updateTransactionSignatures - Function.
 * This function updates transaction object to include the signatures at correct index.
 * The below logic is copied from Lisk SDK https://github.com/LiskHQ/lisk-sdk/blob/2593d1fe70154a9209b713994a252c494cad7123/elements/lisk-transactions/src/sign.ts#L228-L297
 */
/* eslint-disable max-statements, complexity */
const updateTransactionSignatures = (
  account,
  transaction,
  signature,
) => {
  const isMultisigReg = transaction.module === 4; // @todo fix this
  const signerPublicKey = Buffer.from(account.summary.publicKey, 'hex');
  const isSender = Buffer.isBuffer(transaction.senderPublicKey)
    && signerPublicKey.equals(transaction.senderPublicKey);
  const { mandatoryKeys, optionalKeys } = {}; // @todo get this keys from sender account
  if (
    mandatoryKeys.length + optionalKeys.length === 0
    || (isSender && isMultisigReg)
  ) {
    transaction.signatures[0] = signature;
  }

  if (mandatoryKeys.length + optionalKeys.length > 0) {
    const mandatoryKeyIndex = mandatoryKeys.findIndex(
      aPublicKey => isKeyMatch(aPublicKey, signerPublicKey),
    );
    const optionalKeyIndex = optionalKeys.findIndex(
      aPublicKey => isKeyMatch(aPublicKey, signerPublicKey),
    );
    const signatureOffset = isMultisigReg ? 1 : 0;
    if (mandatoryKeyIndex !== -1) {
      transaction.signatures[mandatoryKeyIndex + signatureOffset] = signature;
    }
    if (optionalKeyIndex !== -1) {
      const index = mandatoryKeys.length + optionalKeyIndex + signatureOffset;
      transaction.signatures[index] = signature;
    }
    const numberOfSignatures = signatureOffset + mandatoryKeys.length + optionalKeys.length;
    for (let i = 0; i < numberOfSignatures; i += 1) {
      if (Array.isArray(transaction.signatures)
        && transaction.signatures[i] === undefined) {
        transaction.signatures[i] = Buffer.alloc(0);
      }
    }
  }

  return transaction;
};
/* eslint-disable max-statements */

const signMessageByHW = async ({
  wallet,
  message,
}) => {
  try {
    const signature = await signMessage({
      deviceId: wallet.hwInfo.deviceId,
      index: wallet.hwInfo.derivationIndex,
      message,
    });

    if (!signature) {
      throw new Error(i18next.t(
        'The message signature has been canceled on your {{model}}',
        { model: wallet.hwInfo.deviceModel },
      ));
    }

    return signature;
  } catch (error) {
    throw new Error(i18next.t(
      'The message signature has been canceled on your {{model}}',
      { model: wallet.hwInfo.deviceModel },
    ));
  }
};

export const getDeviceType = (deviceModel = '') => {
  if (/ledger/i.test(deviceModel)) {
    return 'ledgerNano';
  }
  if (/trezor/i.test(deviceModel)) {
    return 'trezor';
  }
  return '';
};

export {
  getNewAccountByIndex,
  checkIfInsideLiskApp,
  getAccountsFromDevice,
  getAddress,
  getPublicKey,
  updateTransactionSignatures,
  subscribeToDeviceConnected,
  subscribeToDeviceDisconnected,
  subscribeToDevicesList,
  validatePin,
  signMessageByHW,
};
