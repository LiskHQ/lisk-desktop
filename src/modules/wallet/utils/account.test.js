import { getCustomDerivationKeyPair } from 'src/utils/explicitBipKeyDerivation';
import accounts from '@tests/constants/wallets';
import {
  extractPublicKey,
  extractPrivateKey,
  extractAddressFromPublicKey,
  calculateUnlockableBalance,
  getUnlockableUnlockObjects,
  calculateBalanceLockedInVotes,
  extractAddressFromPassphrase,
  calculateRemainingAndSignedMembers,
} from './account';

const passphrase = accounts.genesis.passphrase;
const {
  address,
  privateKey,
  publicKey,
} = accounts.genesis.summary;

jest.mock('src/utils/explicitBipKeyDerivation', () => ({
  getCustomDerivationKeyPair: jest.fn(),
}));

describe.skip('Utils: Account', () => {
  describe('extractPublicKey', () => {
    it('should return a hex string from any given string', () => {
      expect(extractPublicKey(passphrase)).toEqual(publicKey);
    });

    it('should call getCustomDerivationKeyPair', () => {
      extractPublicKey(passphrase, true, '1/2');
      expect(getCustomDerivationKeyPair).toHaveBeenCalledWith(passphrase, '1/2');
    });
  });

  describe('extractPrivateKey', () => {
    it('should return a hex string from any given string', () => {
      expect(extractPrivateKey(passphrase)).toEqual(privateKey);
    });

    it('should call getCustomDerivationKeyPair', () => {
      extractPrivateKey(passphrase, true, '1/2');
      expect(getCustomDerivationKeyPair).toHaveBeenCalledWith(passphrase, '1/2');
    });
  });

  describe('extractAddressFromPublicKey', () => {
    it('should return the address corresponding to a (hex) public key', () => {
      expect(extractAddressFromPublicKey(publicKey)).toEqual(address);
    });

    it('should return the address corresponding to a (binary) public key', () => {
      const binaryPublicKey = Buffer.from(publicKey, 'hex');
      expect(extractAddressFromPublicKey(binaryPublicKey)).toEqual(address);
    });
  });

  describe('extractAddressFromPassphrase', () => {
    it('should return the address corresponding to a passphrase', () => {
      expect(extractAddressFromPassphrase(passphrase)).toEqual(address);
    });
  });

  describe('unlocking util functions', () => {
    it('should get correct available balance', () => {
      let unlocking = [
        { amount: '1000000000', height: { start: 4900, end: 5900 }, delegateAddress: 'lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y11' },
        { amount: '3000000000', height: { start: 100, end: 200 }, delegateAddress: 'lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y11' },
        { amount: '1000000000', height: { start: 3000, end: 4000 }, delegateAddress: 'lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y13' },
      ];
      const delegateAddress = '80L';
      const currentBlockHeight = 5000;

      expect(
        calculateUnlockableBalance(unlocking, currentBlockHeight),
      ).toEqual(4000000000);

      unlocking = [
        { amount: '1000000000', height: { start: 4900, end: 5900 }, delegateAddress: 'lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y11' },
        { amount: '3000000000', height: { start: 2500, end: 5500 }, delegateAddress },
        { amount: '1000000000', height: { start: 3000, end: 5500 }, delegateAddress: 'lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y13' },
      ];
      expect(
        calculateUnlockableBalance(unlocking, currentBlockHeight),
      ).toEqual(0);
    });

    it('should return 0 when unlocking is undefined', () => {
      const currentBlockHeight = 5000;
      expect(
        calculateUnlockableBalance(undefined, currentBlockHeight),
      ).toEqual(0);
    });

    describe('calculateBalanceLockedInVotes', () => {
      it('should get correct available balance', () => {
        const votes = {
          lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y11: { confirmed: 5000000000 },
          lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y12: { confirmed: 3000000000 },
          lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y13: { confirmed: 2000000000 },
        };

        expect(calculateBalanceLockedInVotes(votes)).toEqual(10000000000);
      });

      it('should return 0 when unlocking is undefined', () => {
        expect(calculateBalanceLockedInVotes({})).toEqual(0);
      });
    });

    describe('getAvailableUnlockingTransactions', () => {
      it('should get correct available balance', () => {
        const unlocking = [
          { amount: '1000000000', height: { start: 5000, end: 6000 }, delegateAddress: 'lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y11' },
          { amount: '3000000000', height: { start: 100, end: 2000 }, delegateAddress: 'lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y11' },
          { amount: '1000000000', height: { start: 3100, end: 41000 }, delegateAddress: 'lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y13' },
        ];
        const currentBlockHeight = 5000;

        expect(
          getUnlockableUnlockObjects(unlocking, currentBlockHeight),
        ).toEqual([{ amount: '3000000000', unvoteHeight: 100, delegateAddress: 'lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y11' }]);
      });

      it('should return 0 when unlocking is undefined', () => {
        const currentBlockHeight = 5000;
        expect(getUnlockableUnlockObjects(undefined, currentBlockHeight)).toEqual([]);
      });
    });
  });

  describe('calculateRemainingAndSignedMembers', () => {
    it('should return signed and remaining members', () => {
      const keys = {
        mandatoryKeys: ['c5e64031407c3ca8d526bf7404f7c78ab60ea0792e90393a73b3b06a8c8841d4',
          'a1fa251b368939ed2aa8c620e955cb4537c06a351fa50e928ec21e89372e7494'],
        optionalKeys: ['35c6b25520fc868b56c83fed6e1c89bb350fb7994a5da0bcea7a4f621f948c7f',
          '0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a',
          '86499879448d1b0215d59cbf078836e3d7d9d2782d56a2274a568761bff36f19'],
      };

      const signatures = [];
      const result = calculateRemainingAndSignedMembers(keys, signatures);
      expect(result).toMatchSnapshot();
    });

    it('should return signed and remaining members', () => {
      const keys = {
        mandatoryKeys: ['c5e64031407c3ca8d526bf7404f7c78ab60ea0792e90393a73b3b06a8c8841d4',
          'a1fa251b368939ed2aa8c620e955cb4537c06a351fa50e928ec21e89372e7494'],
        optionalKeys: ['35c6b25520fc868b56c83fed6e1c89bb350fb7994a5da0bcea7a4f621f948c7f',
          '0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a',
          '86499879448d1b0215d59cbf078836e3d7d9d2782d56a2274a568761bff36f19'],
      };

      const signatures = ['haha'];
      const result = calculateRemainingAndSignedMembers(keys, signatures);
      expect(result).toMatchSnapshot();
    });

    it('should return signed and remaining members', () => {
      const keys = {
        mandatoryKeys: ['c5e64031407c3ca8d526bf7404f7c78ab60ea0792e90393a73b3b06a8c8841d4',
          'a1fa251b368939ed2aa8c620e955cb4537c06a351fa50e928ec21e89372e7494'],
        optionalKeys: ['35c6b25520fc868b56c83fed6e1c89bb350fb7994a5da0bcea7a4f621f948c7f',
          '0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a',
          '86499879448d1b0215d59cbf078836e3d7d9d2782d56a2274a568761bff36f19'],
      };

      const signatures = [
        'sender_signature',
        'haha',
        '',
        'does not matter',
      ];
      const result = calculateRemainingAndSignedMembers(keys, signatures, true);
      expect(result).toMatchSnapshot();
    });
  });
});
