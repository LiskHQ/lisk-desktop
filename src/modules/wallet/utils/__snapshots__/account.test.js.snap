// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils: Account calculateRemainingAndSignedMembers should return signed and remaining members 1`] = `
Object {
  "remaining": Array [
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": true,
      "publicKey": "c5e64031407c3ca8d526bf7404f7c78ab60ea0792e90393a73b3b06a8c8841d4",
    },
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": true,
      "publicKey": "a1fa251b368939ed2aa8c620e955cb4537c06a351fa50e928ec21e89372e7494",
    },
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": false,
      "publicKey": "35c6b25520fc868b56c83fed6e1c89bb350fb7994a5da0bcea7a4f621f948c7f",
    },
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": false,
      "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
    },
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": false,
      "publicKey": "86499879448d1b0215d59cbf078836e3d7d9d2782d56a2274a568761bff36f19",
    },
  ],
  "signed": Array [],
}
`;

exports[`Utils: Account calculateRemainingAndSignedMembers should return signed and remaining members 2`] = `
Object {
  "remaining": Array [
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": true,
      "publicKey": "a1fa251b368939ed2aa8c620e955cb4537c06a351fa50e928ec21e89372e7494",
    },
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": false,
      "publicKey": "35c6b25520fc868b56c83fed6e1c89bb350fb7994a5da0bcea7a4f621f948c7f",
    },
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": false,
      "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
    },
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": false,
      "publicKey": "86499879448d1b0215d59cbf078836e3d7d9d2782d56a2274a568761bff36f19",
    },
  ],
  "signed": Array [
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": true,
      "publicKey": "c5e64031407c3ca8d526bf7404f7c78ab60ea0792e90393a73b3b06a8c8841d4",
    },
  ],
}
`;

exports[`Utils: Account calculateRemainingAndSignedMembers should return signed and remaining members 3`] = `
Object {
  "remaining": Array [
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": false,
      "publicKey": "35c6b25520fc868b56c83fed6e1c89bb350fb7994a5da0bcea7a4f621f948c7f",
    },
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": false,
      "publicKey": "86499879448d1b0215d59cbf078836e3d7d9d2782d56a2274a568761bff36f19",
    },
  ],
  "signed": Array [
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": true,
      "publicKey": "c5e64031407c3ca8d526bf7404f7c78ab60ea0792e90393a73b3b06a8c8841d4",
    },
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": true,
      "publicKey": "a1fa251b368939ed2aa8c620e955cb4537c06a351fa50e928ec21e89372e7494",
    },
    Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
      "mandatory": false,
      "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
    },
  ],
}
`;
